src/rust/catloop/futures/accept.rs

```diff
@@ -0,0 +1,214 @@
+
+// Checks if a connection request is valid.
+fn check_connect_request(catmem: &Rc<RefCell<CatmemLibOS>>, handle: SchedulerHandle, qt: QToken) -> Result<(), Fail> {
+    // Parse and check request.
+    let passed: bool = {
+        let qr: demi_qresult_t = catmem.borrow_mut().pack_result(handle, qt)?;
+        let sga: demi_sgarray_t = match qr.qr_opcode {
+            demi_opcode_t::DEMI_OPC_POP => unsafe { qr.qr_value.sga },
+            _ => panic!("unxpected operation on control duplex pipe"),
+        };
+        let passed: bool = CatloopLibOS::is_magic_connect(&sga);
+        catmem.borrow_mut().free_sgarray(sga)?;
+        passed
+    };
+
+    if !passed {
+        let e: Fail = Fail::new(libc::EPROTO, "invalid request");
+        error!("failed to establish connection ({:?})", e);
+        Err(e)
+    } else {
+        Ok(())
+    }
+}
+
+// Sends the port number to the peer process.
+fn send_port_number(
+    catmem: &Rc<RefCell<CatmemLibOS>>,
+    control_duplex_pipe: Rc<DuplexPipe>,
+    port: u16,
+) -> Result<QToken, Fail> {
+    let sga: demi_sgarray_t = catmem.borrow_mut().alloc_sgarray(mem::size_of_val(&port))?;
+    let ptr: *mut u8 = sga.sga_segs[0].sgaseg_buf as *mut u8;
+    let len: usize = sga.sga_segs[0].sgaseg_len as usize;
+    let slice: &mut [u8] = unsafe { slice::from_raw_parts_mut(ptr, len) };
+    slice.copy_from_slice(&port.to_ne_bytes());
+    let qt_tx: QToken = control_duplex_pipe.push(&sga)?;
+    catmem.borrow_mut().free_sgarray(sga)?;
+    Ok(qt_tx)
+}
+
+/// Waits for a connection request to arrive.
+fn listen_and_accept_connections(
+    self_: &mut AcceptFuture,
+    ctx: &mut Context<'_>,
+    qt_rx: QToken,
+) -> Poll<Result<(QDesc, SocketAddrV4, Rc<DuplexPipe>), Fail>> {
+    // Check if a connection request arrived.
+    if let Some(handle) = DuplexPipe::poll(&self_.catmem, qt_rx)? {
+        // Check if this is a valid connection request.
+        check_connect_request(&self_.catmem, handle, qt_rx)?;
+
+        // Create underlying pipes before sending the port number through the
+        // control duplex pipe. This prevents us from running into a race
+        // condition were the remote makes progress faster than us and attempts
+        // to open the duplex pipe before that it is created.
+        let duplex_pipe: Rc<DuplexPipe> = Rc::new(DuplexPipe::create_duplex_pipe(
+            self_.catmem.clone(),
+            &self_.ipv4,
+            self_.new_port,
+        )?);
+
+        // Send port number.
+        let qt_tx: QToken = send_port_number(&self_.catmem, self_.control_duplex_pipe.clone(), self_.new_port)?;
+
+        // Advance to next state in the connection establishment protocol.
+        self_.state = ServerState::AcceptedConnectRequest(qt_tx, duplex_pipe.clone());
+    }
+
+    // Re-schedule co-routine for later execution.
+    ctx.waker().wake_by_ref();
+    return Poll::Pending;
+}
+
+fn establish_connection(
```

Comment:
Can we name this just `connect`? Or would that hide some implied meaning?

Reply:
Sure, I renamed it.